File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.ts
function lengthOfLongestSubstring(s: string): number {
    let l = 0;
    const mp = new Map<string,number>();
    let curMax = 0;
    for(let r = 0; r < s.length; r ++ ) {
        if(mp.has(s.charAt(r))) l = Math.max(l,mp.get(s.charAt(r)) + 1);
        mp.set(s.charAt(r),r);
        // console.log(r , l);
        curMax = Math.max(curMax,r - l + 1 );
    }
    return curMax;
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character , Integer> mp = new HashMap<>();
        int l = 0;
        int maxLen = 0;
        for(int r = 0; r < s.length(); r ++) {
            char curChar = s.charAt(r);
            if(mp.containsKey(curChar)){
                int prevIndex = mp.get(curChar);
                l = Math.max(l, prevIndex + 1);
            }
            mp.put(curChar , r);
            int curLen = r - l + 1;
            maxLen = Math.max(maxLen , curLen);
        }
        return maxLen;
    }
}